src/Tizen.System.Session/Session/Session.cs

```diff
@@ -0,0 +1,509 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Tizen.System
+{
+    /// <summary>
+    /// Provides methods to manage subsession users. Allows to register for events triggered by operations on subsession users.
+    /// </summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public sealed class Session
+    {
+        /// <summary>
+        /// UID of current system user.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public static readonly int CurrentUID;
+
+        /// <summary>
+        /// Maximum length of any given user ID.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public const int MaxUserLength = 20;
+
+        /// <summary>
+        /// Special subsession ID, which is always present and does not represent any user.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public const string EmptyUser = "";
+
+        private static ConcurrentDictionary<int, Session> s_sessionInstances = new ConcurrentDictionary<int, Session>();
+
+        private readonly object _replyLock = new object();
+
+        private IDictionary<int, Interop.Session.SubsessionReplyCallback> _replyMap = new ConcurrentDictionary<int, Interop.Session.SubsessionReplyCallback>();
+
+        private int _replyID = 0;
+
+        private delegate void EventDelegate(SubsessionEventInfoNative infoNative, IntPtr data);
+
+        static Session()
+        {
+            CurrentUID = Interop.Session.GetUID();
+        }
+
+        private Session(int sessionUID)
+        {
+            SessionUID = sessionUID;
+        }
+
+        /// <summary>
+        /// Gets a Session object instance for a given sessionUID.
+        /// </summary>
+        /// <param name="sessionUID">Session UID of a requested Session object instance</param>
+        /// <returns>Returns requested Session object</returns>
+        /// <remarks>
+        /// To ensure thread safety, expilicit creation of Session object is not allowed.
+        /// </remarks>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public static Session GetInstance(int sessionUID)
+        {
+            s_sessionInstances.TryAdd(sessionUID, new Session(sessionUID));
```

Comment:
It seems to wrong, 
It will temperory create Session always
```suggestion
            if (!s_sessionInstances.ContainsKey(sessionUID))
                  s_sessionInstances.TryAdd(sessionUID, new Session(sessionUID));
```